waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,157 @@
+  return $
+    testGroup "Wasp.LSP.Completion.getCompletionsAtPosition" $
+      makeGoldenCompletionTest <$> completionTestInputFiles
+
+-- | Create a test case from a pair of <test_name>.wasp and <test_name>.golden
+-- file, using 'runCompletionTest' on the .wasp file to get the output to
+-- compare with or store in the .golden file.
+--
+-- The given filepath should be to a completion test input file (ending in .wasp)
+-- inside the completionTests folder.
+makeGoldenCompletionTest :: FilePath -> TestTree
+makeGoldenCompletionTest inputFp =
+  let goldenFile = replaceExtension inputFp ".golden"
+      testCaseName = takeBaseName inputFp
+      diffCmd = \ref new -> ["diff", "-u", ref, new]
+   in goldenVsStringDiff
+        testCaseName
+        diffCmd
+        goldenFile
+        ( do
+            source <- BSC.unpack <$> BS.readFile inputFp
+            return $ BSC.pack $ runCompletionTest $ CompletionTestInput source
+        )
+
+-- | Takes a completion test input and produces the list of completion items,
+-- represented as a string so it can be compared with the contents of and stored
+-- in the golden file associated with the test input.
+runCompletionTest :: CompletionTestInput -> String
+runCompletionTest testInput =
+  let (waspSource, cursorPosition) = parseCompletionInput testInput
+      tokens = Lexer.lex waspSource
+      parsedCST = snd $ parseCST tokens
+      serverState =
+        ServerState
+          { _currentWaspSource = waspSource,
+            _latestDiagnostics = [],
+            _cst = Just parsedCST
+          }
+      (completionItems, _log) = runLog $ evalStateT (getCompletionsAtPosition cursorPosition) serverState
+      fmtedCompletionItems = map fmtCompletionItem completionItems
+
+      fmtCompletionItem :: LSP.CompletionItem -> String
+      fmtCompletionItem item =
+        unwords
+          [ printf "label={%s}" (show $ item ^. LSP.label),
+            printf "kind={%s}" (show $ item ^. LSP.kind),
+            printf "detail={%s}" (show $ item ^. LSP.detail)
+          ]
+   in "Completion items:\n" ++ unlines (map ("  " <>) fmtedCompletionItems)
+
+-- | Parses a completion test case into a pair of the wasp source code to
+-- run completion on and the position to get the completion list at.
+--
+-- See 'CompletionTestInput' for the format of the test input.
+parseCompletionInput :: CompletionTestInput -> (String, LSP.Position)
+parseCompletionInput (CompletionTestInput testInput) =
+  withPreambleAssert "//! test/completion" testInput (unlines code, position)
+  where
+    -- Drops the marked line AND the following line (which is blank except for the ^)
+    code = before markedLineIdx (lines testInput) ++ [markedLine] ++ after (markedLineIdx + 1) (lines testInput)
+    position = LSP.Position (fromIntegral markedLineIdx) (fromIntegral markedColIdx)
+
+    markedLine = before markedColIdx rawMarkedLine ++ after markedColIdx rawMarkedLine
```

Comment:
"marked" line semantic confused me somewhat. I would maybe go for "lineWithCursor" and then also have function `lineWithCursorCleanedUp` or something like that.